src/utils.cpp

```diff
@@ -195,6 +199,128 @@ bool is_lit(const Term & l, const Sort & boolsort)
   return false;
 }
 
+void cnf_to_dimacs(Term cnf, std::ostringstream & y)
+{
+  if (cnf->is_value() && cnf->to_string() == "true")
+  {  // empty cnf formula
+    y << "p cnf 0 0\n";
+    return;
+  }
+  TermVec vecs({ cnf });
+  TermVec vecs2;
+  // This while loop separate the clauses, vecs2 will contain all clauses
+  // because every smt::and op will be eliminated. This happens because until
+  // smt::and op is not detected that term is added back to vecs, and as no term
+  // with smt::or as the primOp is touched all clauses shall be separated and be
+  // intact
+  while (!vecs.empty())
+  {
+    Term t = vecs.back();
+    vecs.pop_back();
+    smt::Op op = t->get_op();
+    if (op.prim_op == smt::And)
+    {
+      for (auto u : t)
+      {
+        vecs.push_back(u);
+      }
+    }
+    else
+    {
+      vecs2.push_back(t);
+    }
+  }
+  // Storing literals from each clause, each vector in literals will contain the
+  // literals from a clause
+  std::vector<std::vector<Term>> literals;
+
+  for (auto u : vecs2)
+  {
+    std::vector<Term>add;
+    std::vector<Term>le({u});
+    while (!le.empty())
+    {  // This while loop functions in the same way as above and eliminates
+       // smt::or by separating the literals
+      Term t=le.back();
+      le.pop_back();
+      smt::Op op=t->get_op();
+      if(op.prim_op==smt::Or){
+        for(auto u:t){
+          le.push_back(u);
+        }
+      }
+      else{
+        add.push_back(t);
+      }
+    }
+    literals.push_back(add);
+   }
+
+   std::map<std::string, int>
```

Comment:
Why from `std::string` to `int` and not from `Term` to `int`?

Reply:
I had a feeling that mapping std::string would lead to less overhead than 'Term', though I might be wrong. Is there any disadvantage in using std::string?

Reply:
Since the strings are just variable names, I don't think there will be a noticeable difference in the overhead. But even if there is, I expect the overhead of storing terms will be smaller, because probably the most naive hashing function that solvers can implement is transforming to a string and then compute the hash. But some solvers do much better than that.

The disadvantage I see in using `string` is the extra level of indirection. I think what we are really interested in this map are the terms, their string representation is somewhat irrelevant, unless I am missing its relevancy.

Reply:
Ok, I'll change this

Reply:
Btw I had first thought of just entering the terms in a vector, sorting them and them using binary_search on the terms to find their index, but it seemed like this function would be the least expensive computationally and mapping seemed the shortest to code